### PR TITLE
Pks

### PR DESCRIPTION
**Pull request by JaimeRZP**

### CODE REVIEW

File: src/halofit.jl [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    neff2 = @. neff * neff
+    neff3 = @. neff2 * neff
+    neff4 = @. neff3 * neff
+
+    delta2_norm = @. k*k*k/2.0/pi/pi
+
+    # compute the present day neutrino massive neutrino fraction
+    # uses all neutrinos even if they are moving fast
+#     om_nu = cosmo.sum_nu_masses / 93.14 / cosmo.h / cosmo.h
+#     fnu = om_nu / (cosmo.Ωm)
```
Comment (damonge) [inline src/halofit.jl:152]: What happened here?

**Review by damonge — APPROVED** [review]
OK, mostly looks good.

File: src/core.jl [hunk 2]
```diff
@@ -194,6 +192,20 @@ function TkEisHu(cosmo::CosmoPar, k)
...
 
+function _omega_x(cosmo::CosmoPar, z, species_x_label)
+    Ez = _Ez(cosmo, z)
+    a = @. 1.0/(1.0 + z)
+    if species_x_label == "crit"
+        return 1.0
+    elseif species_x_label == "m"
+        return @. cosmo.Ωm / (a^3) / Ez^2
+    elseif species_x_label == "l"
+        return 1.0 - cosmo.Ωm
```
[the remark this answers is not in the document]
Comment (JaimeRZP) [reply src/core.jl]: I have removed these lines from core and moved them to Halofit.
If they are wrong here, they are also wrong there.

File: src/core.jl [hunk 3]
```diff
@@ -209,7 +221,7 @@ Ez(cosmo::Cosmology, z) = _Ez(cosmo.cosmo, z)
 Hmpc(cosmo::Cosmology, z) = cosmo.cosmo.h*Ez(cosmo, z)/CLIGHT_HMPC
 comoving_radial_distance(cosmo::Cosmology, z) = cosmo.chi(z)
 growth_factor(cosmo::Cosmology, z) = cosmo.Dz(z)
+omega_x(cosmo::Cosmology, z, species_x_label) = _omega_x(cosmo.cosmo, z, species_x_label)
 function power_spectrum(cosmo::Cosmology, k, z)
-    Dz2 = growth_factor(cosmo, z)^2
-    @. exp(cosmo.lplk(log(k)))*Dz2
-end
+    @. exp(cosmo.lplk(log(k)))*cosmo.Dz(z)*cosmo.Dz(z)
```
[the remark this answers is not in the document]
Comment (JaimeRZP) [reply src/core.jl]: Done!

File: src/halofit.jl [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    z = range(zmin, stop=zmax, length=nz)
+    a = reverse(@. 1.0 / (1.0 + z))
+    zs = reverse(z)
+    
+    #=
+    PkL = zeros(length(a), length(k))
+    for i in range(1, stop=length(a))
+        PkL[i, :] .= power_spectrum(cosmo, k, zs[i])
+    end
+    =#
```
[the remark this answers is not in the document]
Comment (JaimeRZP) [reply src/halofit.jl]: Done!

File: src/halofit.jl [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    result = trapz(logk, gauss_norm_int_func(logk, pk, rsigma)) - 1.0
+    return result
+end
+
+function solve_for_rsigma(pk, logk)
+    rlow = 1e-2
+    rhigh = 1e2
+#     f1 = rsig -> rsigma_func(rsig, pk, logk)
+#     f2 = rsig -> trapz(logk, onederiv_gauss_norm_int_func(logk, pk, rsig))
+#     rsigma = find_zero((f1, f2), 5.5, Roots.Newton())
```
[the remark this answers is not in the document]
Comment (JaimeRZP) [reply src/halofit.jl:83]: No, these were commented by Andrina. 
We could ask her.

File: src/halofit.jl [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+    zs = @. 1.0/a - 1.0
+    weffa = -1.0
+    omegaMz = omega_x(cosmo, zs, "m")
+    omegaDEwz = omega_x(cosmo, zs,"l")
+
+    # not using these to match CLASS better - might be a bug in CLASS
+    # weffa = gsl_spline_eval(hf->weff, a, NULL);
+    # omegaMz = gsl_spline_eval(hf->omeff, a, NULL);
+    # omegaDEwz = gsl_spline_eval(hf->deeff, a, NULL);
```
[the remark this answers is not in the document]
Comment (JaimeRZP) [reply src/halofit.jl:128]: No, these were commented by Andrina. 
We could ask her.

File: src/tracers.jl [hunk 7]
```diff
@@ -82,4 +82,4 @@ function get_Fℓ(t::Tracer, ℓ)
     else
         return 1
     end
-end
+end
```
[the remark this answers is not in the document]
Comment (JaimeRZP) [reply src/tracers.jl:85]: Done!

File: src/halofit.jl (hunk 1 above)
Comment (JaimeRZP) [reply src/halofit.jl:140]: These were commented by Andrina. 
We could ask her.

File: src/core.jl [hunk 8]
```diff
@@ -209,7 +221,26 @@ Ez(cosmo::Cosmology, z) = _Ez(cosmo.cosmo, z)
...
-function power_spectrum(cosmo::Cosmology, k, z)
-    Dz2 = growth_factor(cosmo, z)^2
-    @. exp(cosmo.lplk(log(k)))*Dz2
+omega_x(cosmo::Cosmology, z, species_x_label) = _omega_x(cosmo.cosmo, z, species_x_label)
+
+#function power_spectrum(cosmo::Cosmology, k, z)
+    # @. exp(cosmo.primordial_lPk(log(k)))*cosmo.Dz(z)*cosmo.Dz(z)
+    #PKls, Pk_itp = cosmo.Pk(k, z)
+#    return cosmo.PkL(k, z)
+#end
```
Comment (damonge) [inline src/core.jl:230]: Remove these

File: src/core.jl [hunk 9]
```diff
@@ -209,7 +221,26 @@ Ez(cosmo::Cosmology, z) = _Ez(cosmo.cosmo, z)
 Hmpc(cosmo::Cosmology, z) = cosmo.cosmo.h*Ez(cosmo, z)/CLIGHT_HMPC
 comoving_radial_distance(cosmo::Cosmology, z) = cosmo.chi(z)
 growth_factor(cosmo::Cosmology, z) = cosmo.Dz(z)
-function power_spectrum(cosmo::Cosmology, k, z)
-    Dz2 = growth_factor(cosmo, z)^2
-    @. exp(cosmo.lplk(log(k)))*Dz2
+omega_x(cosmo::Cosmology, z, species_x_label) = _omega_x(cosmo.cosmo, z, species_x_label)
```
Comment (damonge) [inline src/core.jl:224]: Where did the `_omega_x` function go? I can't see it anywhere

**Review by damonge — COMMENTED** [review]
A few more comments @JaimeRZP

File: src/core.jl (hunk 8 above)
Comment (JaimeRZP) [reply src/core.jl:230]: Done!

File: src/core.jl (hunk 9 above)
Comment (JaimeRZP) [reply src/core.jl:224]: https://github.com/damonge/LimberJack.jl/blob/c1fbc8d1ed06734eb2182d1f86b7d7ab01033188/src/halofit.jl#L119-L123

Comment (damonge) [reply src/core.jl:224]: ok, then:
1. If `_omega_x` does not exist, this line should disappear. I think there's a similar one in `LimberJack.jl`.
2. The implementation in halofit.jl is wrong (confirmed with Andrina). It should read
```
    omegaDEwz = (1 - cpar.Ωm)/Ez^2
```

Comment (JaimeRZP) [reply src/core.jl]: Ok great! implemented!

File: src/core.jl [hunk 10]
```diff
@@ -209,7 +221,23 @@ Ez(cosmo::Cosmology, z) = _Ez(cosmo.cosmo, z)
...
-    @. exp(cosmo.lplk(log(k)))*Dz2
+
+function nonlin_Pk(cosmo::Cosmology, k, z)
+    return cosmo.Pk(k, z)
 end
+
+function _nonlin_Pk(cpar::CosmoPar, zs, ks, PkLs)
+    Pk = PKnonlin(cpar, zs, ks, PkLs).pk_NL
+    return Pk
+end
```
Comment (damonge) [inline src/core.jl:232]: This function seems redundant. Couldn't you just call `PKnonlin` above?

File: src/core.jl [hunk 11]
```diff
@@ -209,7 +221,23 @@ Ez(cosmo::Cosmology, z) = _Ez(cosmo.cosmo, z)
...
+function lin_Pk(cosmo::Cosmology, k, z)
+    return cosmo.PkL(k, z)
+end
+
+function _lin_Pk(primordial_lPk, Dz, zs, ks)
+    PkLs = [@. exp(primordial_lPk(log(k)))*Dz(zs)^2 for k in ks]
+    PkLs = reduce(vcat, transpose.(PkLs))
+    PkL = LinearInterpolation((ks, zs), PkLs)
+    return PkLs, PkL
+end
```
Comment (damonge) [inline src/core.jl:243]: add a newline

File: src/spectra.jl [hunk 12]
```diff
@@ -11,7 +11,7 @@ function Cℓintegrand(cosmo::Cosmology,
     z = cosmo.z_of_chi(chi)
     w1 = t1.wint(chi)*t1.bias
     w2 = t2.wint(chi)*t2.bias
-    pk = power_spectrum(cosmo, k, z)
+    pk = cosmo.Pk(k, z)
```
Comment (damonge) [inline src/spectra.jl:14]: So this will be always the non-linear power spectrum?

File: src/core.jl [hunk 13]
```diff
@@ -51,7 +51,7 @@ struct Cosmology
     ks::Array
     pk0::Array
     dlogk
-    lplk::AbstractInterpolation
+    primordial_lPk::AbstractInterpolation
```
Comment (damonge) [inline src/core.jl:54]: I'm not sure I understand what you're calling "primordial_lPk" here. I think it is the linear power spectrum at z=0, right? If so, that's not the primordial power spectrum, so I'd call it `linear_lPk_z0`, for instance.

Although, not sure if you really need this, if you're already storing the linear power spectrum at all redshifts as well.

**Review by damonge — CHANGES_REQUESTED** [review]
OK, a few more small comments. Note that you have a conflict with the main branch on `runtests.jl`, so you'll need to fix that one before you can merge

**Review by damonge — APPROVED** [review]
LGTM. Check tests and merge!

**Pull request merged**